doc/architectural/goto-program-transformations.md

```diff
@@ -0,0 +1,249 @@
+`src/pointer-analysis/add_failed_symbols.h` for details.
+
+_Predecessor passes are [Goto Functions Update] or
+[Add Non-Deterministic Initialisation of Global Scoped Variables]._
+
+
+### Remove Skip Instructions
+
+This transformation removes skip instructions. Note that this transformation is
+called in many places and may be called as part of other transformations. This
+instance here is part of the mandatory transformation to reach core goto format.
+
+_Predecessor passes are [Add Failed Symbols] or [Remove Unused Functions]._
+
+
+
+### Label Properties
+
+This transformation adds information to assert instructions. This includes
+- Adding source locations to asserts where they are missing
+- Populating the locations's property ID
+
+_Predecessor passes are [Remove Skip Instructions] or [Add Coverage Goals]._
+
+
+
+## Optional Transformation Passes
+
+The sections lists the optional transformation passes that are optional and will
+modify a goto model. Note that these are documented here for consistency, but
+not require for core goto format.
+
+Note for each optional pass there is a listed predeceesor pass. This is the pass
+currently called before the listed pass in CBMC. While the ordering may not be
+fully documented, this should be followed.
+
+### Linking of Standard Libraries
+
+This pass links the function defenitions from the Cprover standard library
+models to the current goto model.
+
+_Predecessor pass is [Removal/Lowering of Assembly]._
+
+
+### Partial Inlining
+
+This pass does partial inlining when this option is switched on. Partal inlining
+is inlining of functions either: marked as inline, or smaller than a specified
+limit. For further detail see the implementation XXXXXXX
+
+_Predecessor pass is [Instrument/Remove Preconditions]._
+
+
+### Transform Assertions Assumptions
+ 
+ This pass removes user provided assertions and assumptions when user has opted
+ to do so. Note that this pass removes skip instructions if any other changes
+ are made.
+
+_Predecessor pass is [goto_check_c]._
+
+
+### String Abstraction
+
+This (optional) transformation pass adds tracking of length of C strings. It is
+switched on by the `--string-abstraction` command line option.
+
+_Predecessor pass is [Adjust Float Expressions]._
+
+
+### Add Non-Deterministic Initialisation of Global Scoped Variables
+
+This transformation adds non-deterministic initialisation of global scoped
+variables including static variables. For details see
+`src/goto-instrument/nondet_static.h`. The initialisation code is added to the
+`CPROVER_initialize` function in the goto model.
+
+_Predecessor pass is [Goto Functions Update]._
+
+
+### Remove Unused Functions
+
+This pass removes unused functions from the goto model. In practice this builds
+a collection of all the functions that are potentially called, and then removes
+any function not in this collection.
+
+_Predecessor pass is [Add Failed Symbols]._
+
```

Comment:
This pass triggers function pointer removal when applied using goto-instrument, but may not trigger it when applied programatically. Is this pass sound when all pointers have not been resolved ? Should it have function pointer removal as a precondition ?

Reply:
+1

Reply:
Adding (or improving) a precondition would be out-of-scope for a documentation PR. The remove unused functions pass makes the assumption that `function_call` instructions have their `call_function` value assigned to a symbol expression. This assumption is already checked inside the `to_symbol_expr` function used by the pass.

Calls to function pointers which have not been lowered are `function_call` instructions where the value of `call_function` is a dereference expression. Therefore I would expect that programmatically calling the remove unused functions pass on a program which contains an unlowered function pointer call will result in an invariant violation; thus preventing any kind of analysis sound or otherwise from taking place.